src/transactions/typeTag/typeTagParser.ts

```diff
@@ -0,0 +1,246 @@
+      if (savedPop === undefined) {
+        throw new TypeTagParserError("Failed to parse typeTag, unexpected '>'");
+      }
+
+      // If the expected types don't match the number of commas, then we also fail
+      if (expectedTypes !== curTypes.length) {
+        throw new TypeTagParserError("Failed to parse typeTag, type argument count mismatch");
+      }
+
+      // Add in the new created type, shifting the current types to the inner types
+      const { savedStr, savedTypes, savedExpectedTypes } = savedPop;
+      innerTypes = curTypes;
+      curTypes = savedTypes;
+      currentStr = savedStr;
+      expectedTypes = savedExpectedTypes;
+    } else if (char === ",") {
+      // Comma means we need to start parsing a new tag, push the previous one to the curTypes
+      // Process last type, if there is no type string, then don't parse it
+      if (currentStr.length !== 0) {
+        const newType = parseTypeTagInner(currentStr, innerTypes);
+
+        // parse type tag and push it on the types
+        innerTypes = [];
+        curTypes.push(newType);
+        currentStr = "";
+        expectedTypes += 1;
+      }
+    } else if (isValidWhitespaceCharacter(char)) {
+      // This means we should save what we have and everything else should skip until the next ,
+      let parsedTypeTag = false;
+      if (currentStr.length !== 0) {
+        const newType = parseTypeTagInner(currentStr, innerTypes);
+
+        // parse type tag and push it on the types
+        innerTypes = [];
+        curTypes.push(newType);
+        currentStr = "";
+        parsedTypeTag = true;
+      }
+
+      // Skip ahead on any more whitespace
+      cur = consumeWhitespace(typeStr, cur);
+
+      // The next space MUST be a comma, or a closing > if there was something parsed before
+      // e.g. `u8 u8` is invalid but `u8, u8` is valid
+      const nextChar = typeStr[cur];
+      if (parsedTypeTag && nextChar !== "," && nextChar !== ">") {
+        throw new TypeTagParserError(`Failed to parse typeTag, invalid character after space ${nextChar}`);
+      }
+
+      // eslint-disable-next-line no-continue
+      continue;
+    } else {
+      // Any other characters just append to the current string
+      currentStr += char;
+    }
+
+    cur += 1;
+  }
+
+  // This prevents a missing '>' on type arguments
+  if (saved.length > 0) {
+    throw new TypeTagParserError("Invalid typeTag, no matching '>' for '<'");
+  }
+
+  // This prevents 'u8, u8' as an input
+  if (curTypes.length > 0) {
+    throw new TypeTagParserError("Invalid typeTag, unexpected ','");
+  }
+
+  return parseTypeTagInner(currentStr, innerTypes);
+}
+
+/**
+ * Parses a type tag with internal types associated
+ * @param str
+ * @param types
+ */
+function parseTypeTagInner(str: string, types: Array<TypeTag>): TypeTag {
+  switch (str) {
+    case "bool":
+      if (types.length > 0) {
+        throw new TypeTagParserError("Invalid typeTag, bool should not have a type argument");
```

Comment:
not sure that I understand the error message..?

Reply:
Maybe we make a better name but basically:

This is not a type
```
bool<u8>
```
This is the only type that's a bool
```
bool
```